File: unity_code/Gameplay/Pipe.cs
using System;
using System.Collections.Generic;
using Assets.Scripts.Actions;
using System.Linq;
using UnityEngine;
using Random = UnityEngine.Random;

namespace Assets.Scripts.Gameplay
{
    [RequireComponent(typeof(SpriteRenderer), typeof(Transform), typeof(Animator))]
    public class Pipe : MonoBehaviour, IActable
    {
        private const int maxRotation = 4;
        private int rotationState { get; set; }
        private PipeType type { get; set; }
        private Animator animator;

        private bool isSteamConnected;
        private bool isRocketConnected;

        public bool IsReadyToLaunch
        {
            get { return isSteamConnected && isRocketConnected; }
        }

        public Transform Transform
        {
            get
            {
                return transform;
            }
        }

        public void Kill()
        {
            Destroy(gameObject);
        }

        public Pipe UpPipe { get; set; }
        public Pipe RightPipe { get; set; }
        public Pipe DownPipe { get; set; }
        public Pipe LeftPipe { get; set; }

        public event Action Pressed = () => { };

        public bool IsSideConnected(PipeSide side)
        {
            return PipeConnectionSides.Contains(side);
        }

        public void ConnectSteam()
        {
            if(isSteamConnected)
                return;

            isSteamConnected = true;
            SetTint(Color.yellow);

            if (UpPipe != null )
                UpPipe.ConnectSteam();

            if(RightPipe != null)
                RightPipe.ConnectSteam();

            if(DownPipe != null)
                DownPipe.ConnectSteam();

            if (LeftPipe != null)
                LeftPipe.ConnectSteam();
        }

        public void ConnectRocket()
        {
            if (isRocketConnected)
                return;

            isRocketConnected = true;
            SetTint(!IsReadyToLaunch ? Color.red : Color.green);

            if (UpPipe != null)
                UpPipe.ConnectRocket();

            if (RightPipe != null)
                RightPipe.ConnectRocket();

            if (DownPipe != null)
                DownPipe.ConnectRocket();

            if (LeftPipe != null)
                LeftPipe.ConnectRocket();
        }

        private void SetTint(Color color)
        {
            var spriteRenderer = GetComponent<SpriteRenderer>();
            spriteRenderer.color = color;
        }

        public void Reset()
        {
            UpPipe = null;
            DownPipe = null;
            RightPipe = null;
            LeftPipe = null;

            isSteamConnected = false;
            isRocketConnected = false;

            SetTint(Color.white);
        }

        public PipeSide[] PipeConnectionSides
        {
            get
            {
                var sides = GetInitialConnectionSides();
                RotateConnectionSides(sides);
                return sides;
            }
        }

        private PipeSide[] GetInitialConnectionSides()
        {
            switch (type)
            {
                case PipeType.Sides2Bent:
                    return new[] { PipeSide.Left, PipeSide.Up };
                case PipeType.Sides2Straight:
                    return new[] {PipeSide.Left, PipeSide.Right};
                case PipeType.Sides3:
                    return new[] {PipeSide.Left, PipeSide.Right, PipeSide.Up};
                case PipeType.Sides4:
                    return new[] {PipeSide.Up, PipeSide.Right, PipeSide.Down, PipeSide.Left};
            }

            throw new NotSupportedException("The following pipe type is not supported: " + type);
        }

        private void RotateConnectionSides(PipeSide[] pipeSides)
        {
            for (var i = 0; i < pipeSides.Length; i++)
            {
                var rotatedSide = RotateConnectionSide(pipeSides[i]);
                pipeSides[i] = rotatedSide;
            }
        }

        private PipeSide RotateConnectionSide(PipeSide pipeSide)
        {
            return (PipeSide) (((int)pipeSide + rotationState) % GetEnumSize<PipeSide>());
        }

        private void Awake()
        {
            animator = GetComponent<Animator>();
            SetRandomTile();
            SetInitialRotationState();
        }

        private void SetRandomTile()
        {
            type = GenerateType();
            animator.SetTrigger(type.ToString());
        }

        private void SetInitialRotationState()
        {
            rotationState = GenerateRotationState();
            Rotate (rotationState);
        }

        private void Rotate()
        {
            rotationState = ++rotationState % maxRotation;
            transform.Rotate(Vector3.forward, -90);
        }

        private void Rotate(int times)
        {
            transform.Rotate(Vector3.forward, -90 * times);
        }

        private PipeType GenerateType()
        {
            return (PipeType)Random.Range(0, GetEnumSize<PipeType>());
        }

        private int GenerateRotationState()
        {
            return Random.Range(0, maxRotation);
        }

        private int GetEnumSize<T>()
        {
            return Enum.GetValues(typeof(T)).Length;
        }

        private enum PipeType
        {
            Sides3,
            Sides4,
            Sides2Straight,
            Sides2Bent
        }

        void OnMouseDown()
        {
            Rotate();
            Pressed();
        }
    }
}

File: src/main/ts/model/station/Pipe.ts
import { PipeCreatedEvent } from './events/PipeCreatedEvent';
import { Config } from "../../Config";
import { IActable } from "../actions/IActable";
import { PipeType } from "./data/PipeType";
import { PipeSide } from "./data/PipeSide";
import { IEventDispatcher } from "robotlegs";
import { PipeUpdatedEvent } from './events/PipeUpdatedEvent';

export class Pipe implements IActable {
    private rotationState:number = 0;

    public type:PipeType;
    
    public isSteamConnected:boolean = false;
    public isRocketConnected:boolean = false;

    public x:number;
    public y:number;

    public i:number;
    public j:number;

    public rotation:number;

    public upPipe:Pipe;
    public rightPipe:Pipe;
    public downPipe:Pipe;
    public leftPipe:Pipe;

    private eventDispatcher:IEventDispatcher;

    constructor(x:number, y:number, i:number, j:number, eventDispatcher:IEventDispatcher) {
        this.eventDispatcher = eventDispatcher;
        
        this.type = this.generateType();
        this.i = i;
        this.j = j;

        this.x = x;
        this.y = y;

        this.rotate(this.generateRotation());
    }

    public changeIndexes(i:number, j:number) {
        this.i = i;
        this.j = j;
        
        this.eventDispatcher.dispatchEvent(new PipeUpdatedEvent(this));
    }

    public move(x:number, y:number) {
        this.x = x;
        this.y = y;

        this.eventDispatcher.dispatchEvent(new PipeUpdatedEvent(this));
    }

    public connectSteam() {
        if(this.isSteamConnected) {
            return;
        }

        this.isSteamConnected = true;
        this.eventDispatcher.dispatchEvent(new PipeUpdatedEvent(this));

        if(this.upPipe != null) {
            this.upPipe.connectSteam();
        }

        if(this.rightPipe != null) {
            this.rightPipe.connectSteam();
        }

        if(this.downPipe != null) {
            this.downPipe.connectSteam();
        }

        if(this.leftPipe != null) {
            this.leftPipe.connectSteam();
        }
    }

    public connectRocket() {
        if(this.isRocketConnected) {
            return;
        }

        this.isRocketConnected = true;
        this.eventDispatcher.dispatchEvent(new PipeUpdatedEvent(this));

        if(this.upPipe != null) {
            this.upPipe.connectRocket();
        }

        if(this.rightPipe != null) {
            this.rightPipe.connectRocket();
        }

        if(this.downPipe != null) {
            this.downPipe.connectRocket();
        }

        if(this.leftPipe != null) {
            this.leftPipe.connectRocket();
        }
    }

    public isSideConnected(side:PipeSide):boolean {
        return this.getPipeConnectionSides().indexOf(side) > -1;
    }

    public isReadyToLaunch():boolean {
        return this.isSteamConnected && this.isRocketConnected;
    }

    public getPipeConnectionSides():PipeSide[] {
        let sides:PipeSide[] = this.getInitialConnectionSides();
        return this.rotateConnectionSides(sides);
    }

    public reset() {
        this.upPipe = null;
        this.downPipe = null;
        this.rightPipe = null;
        this.leftPipe = null;

        this.isSteamConnected = false;
        this.isRocketConnected = false;

        this.eventDispatcher.dispatchEvent(new PipeUpdatedEvent(this));
    }

    private getInitialConnectionSides():PipeSide[] {
        switch(this.type) {
            case PipeType.Sides2Bent:
                return [PipeSide.Left, PipeSide.Up];
            case PipeType.Sides2Straight:
                return [PipeSide.Left, PipeSide.Right];
            case PipeType.Sides3:
                return [PipeSide.Left, PipeSide.Down, PipeSide.Right];
            case PipeType.Sides4:
                return [PipeSide.Up, PipeSide.Right, PipeSide.Down, PipeSide.Left];
            //case PipeType.Sides1:
            //    return [PipeSide.Left];
        }
    }

    private rotateConnectionSides(sides:PipeSide[]):PipeSide[] {
        for(let i in sides) {
            let rotatedSide = this.rotateConnectionSide(sides[i]);
            sides[i] = rotatedSide;
        }
        return sides;
    }

    private rotateConnectionSide(side:PipeSide) {
        return (side + this.rotationState) % PipeSide.length;
    }

    private generateType():PipeType {
        return Math.round(Math.random() * (PipeType.length - 1));
    }

    private generateRotation():number {
        return Math.round(Math.random() * (PipeSide.length - 1));
    }

    public rotate(times:number = 1) {
        let oldRotationState:number = this.rotationState;
        let newRotationState:number = (oldRotationState + times) % PipeSide.length;
        let rotationTimes:number = newRotationState - oldRotationState;
        this.rotationState = newRotationState;
        this.rotation = 90 * rotationTimes * Math.PI / 180;
    }
}